### DISCUSSION TITLE
IIT Madras Paradox X Fetch.AI hackathon

### DISCUSSION THREAD

**Discussion by abhifetch** [fetchai/uAgents#368] [Hackathons]
Ask your queries here.

**Comment by AndrewDavidRatnam** [comment]
how to use the correct import for Dialogue ?

**Reply by abhifetch** [reply]
you can use "from uagents.experimental.dialogues import Dialogue, Edge, Node"

Can also look into docs at : [dialogue guide](https://fetch.ai/docs/guides/agents/dialogues) and [dialogue exmaple 1](https://fetch.ai/docs/examples/open-dialogue-chitchat), [dialogue exmaple 2](https://fetch.ai/docs/examples/predefined-dialogue-chitchat)

**Reply by AndrewDavidRatnam** [reply]
i got a notification that dialogue is not experimental while running a few days back that's why

**Reply by abhifetch** [reply]
Yes, the dialogues are experimental feature under heavy development. Backward compatibility is not guaranteed, and updates may cause breaking changes. Each update might introduce significant changes in behaviour or functionality.

can you please tell what are you trying to achieve using dialogues

**Reply by AndrewDavidRatnam** [reply]
I'm trying to create discussion and task assignment capabilities

**Reply by AndrewDavidRatnam** [reply]
Should i just loop over protocols or else to be safe?

**Reply by abhifetch** [reply]
there are multiple things you can do:
1. use chat API which is similar to [deltaV](https://fetch.ai/docs/examples/chat-api-example) and assign tasks (can create your own app using API).
2. Agents listen to messages in sequential order, you can create different message handlers to listen to different type of message and act accordingly
3. You can use any other LLM's API to create discussion and make agents calls wherever you feel necessary to.

**Reply by AndrewDavidRatnam** [reply]
I will use Bureau for agents to talk to each other, that seems more intuitive. Cheers for the quick reply

**Reply by abhifetch** [reply]
Thankyou Andrew, please get back in case you have more queries

**Reply by 23f1002382** [reply]
Sure.

**Comment by abhifetch** [comment]
Just one update, i told in workshop that we use Bureau when multiple agents have to talk to each other but not in application development case. 
The correct thing is Bureau is used if we want to run multiple agents on same port. We can use Bureau in application development as well.

**Comment by gkranthy** [comment]
I want to thank to the @team, for introducing uAgents. I am building an application using these agents with react and python. It feels great to work with these advanced technologies.

Open for Collaboration.

**Comment by mdabucse** [comment]
![image](https://github.com/fetchai/uAgents/assets/131733014/d10bc967-8dae-49ba-9d66-d547ab665556)
I can't able to insert the data into it how to resolve it

**Reply by AndrewDavidRatnam** [reply]
append names and age to appointment list as a tuple, not set a new one everytime
get it
update it
set it
ok ?

**Reply by abhifetch** [reply]
use async with your functions, also everytime you set a value  in storage it gets updated instead of appended. so solution given by @AndrewDavidRatnam is right. Thanks @AndrewDavidRatnam

**Comment by AndrewDavidRatnam** [comment]
Is is faster to use TortoiseDB to read and write data or use ctx.storage? Since increasing context size, then transferring(via decentralized that too), must have more overhead right?

**Reply by abhifetch** [reply]
As per my understanding it should depend on type and volume of data that is to be stored. if you have big volume of data definitely DB like TortoiseDB should be considered. Our agent's storage is only capable of storing JSON data as of now.

**Reply by AndrewDavidRatnam** [reply]
yes, mostly local operation as i'm using bureau, so localDB makes more sense. Cheers for the prompt reply

**Reply by abhifetch** [reply]
no worries you can get back here if you have more issues.

**Comment by mdabucse** [comment]
If i want to build an agent that take some inputs and after getting all the it says booked successfully anyone suggest some ideas

**Reply by AndrewDavidRatnam** [reply]
https://fetch.ai/docs/guides/agents/cleaning-demo

**Reply by AndrewDavidRatnam** [reply]
https://fetch.ai/docs/guides/agents/booking-demo

**Reply by abhifetch** [reply]
above mentioned examples can be used.

**Comment by AndrewDavidRatnam** [comment]
Can my agent, which includes a protocol with some handler with some function with( ctx:Context), 
1)This context is agent's context, protocol has no context right? First Question
2)Can my context storage for said agent hold other agents? Second Question
3)How far down will this go? Is there an error handler to some limited value that agents can hold? if second question is yes?

**Reply by AndrewDavidRatnam** [reply]
I ask this because I need to dynamically create agents, is there process to delete agent etc? or some dynamic garbage collecter to create some type of dynamic agent and clean it up after use?

**Reply by abhifetch** [reply]
1. yes if you call a handler with an protocol and protocol is included in the agent, it is agent's context. 
2. no it is not possible, it can hold addresses if you want to 
3. the answer is no it can't hold other agents. It can store addresses

also if you want to create agents dynamically and delete it you can use API docs here to do so. 
guide : https://fetch.ai/docs/apis/agentverse/hosting
example : https://fetch.ai/docs/examples/agent-and-service-api

**Reply by AndrewDavidRatnam** [reply]
NPC = ctx.storage.get("NPC")
        for element in interactive_elements:
            NPC.append(Agent(name=element.name, seed=element.name)) #Why wont this work? 
            BuildElement(NPC[-1])
        ctx.storage.set("NPC",NPC) #store the NPC in the context
        
     Will this throw an error?

**Reply by abhifetch** [reply]
You are trying to store an object in agent's storage. Agent storage is only capable of storing JSON format.
Object of type Agent is not JSON serializable.

**Reply by AndrewDavidRatnam** [reply]
That makes so much more sense now! Thank you for helping me.

**Comment by 23f1002382** [comment]
Sure, here’s a more conversational and persuasive version suitable for a discussion forum:

---

Hey everyone,

I have a suggestion for improving our agent submission process. Since we can create multiple agents, how about we each create a folder with our IDs or credentials? We can submit pull requests for these folders instead of directly to the main integrations folder. After evaluations and reviews, we can then PR the main folder.

Here's why I think this could help:

1. **Better Organization**: Having a personal folder for each of us keeps things tidy and makes it easier to find and manage our agents.
2. **Easier Reviews**: Reviewers can focus on one folder at a time, which should streamline the review process and reduce the chance of missing something.
3. **Efficient Submissions**: We can handle all our agents in one place, making it easier to update and test them before they're merged into the main project.
4. **Less Congestion**: This keeps the main integrations folder cleaner, only adding agents that have been fully reviewed and approved.

### How It Could Work:

1. **Create Your Folder**: Each of us creates a folder named with our ID or credentials in a parent directory.
2. **Develop and Test**: Work on your agents in your folder, making sure everything is solid.
3. **Submit Initial PR**: When ready, submit a PR for your folder.
4. **Review and Feedback**: Reviewers check out your folder, give feedback, and request changes if needed.
5. **Final Integration**: After approval, we merge the contents into the main integrations folder.

This approach should make our submissions clearer and reviews more efficient. What do you all think?

Cheers!

*GPT-4o

**Reply by AndrewDavidRatnam** [reply]
Makes Sense. Let's see what the others say.

**Reply by abhifetch** [reply]
@ANdIeCOOl feels great that you have provided such great details. Its already shared that how we are expecting the submissions tk be. I will again share the details bY EOD which helps us identify your code and keep our repo clean and organised.

**Reply by 23f1002382** [reply]
NO NEED. https://github.com/fetchai/uAgents/discussions/368#discussioncomment-9554521

**Comment by saquib000** [comment]
Hi! So i am thinking to pivot on my problem statement a little bit. Earlier i was thinking to do something on delta v, but now i thinking to do a separate user-interactive app.
The doubt i am having is, would it be enough if i just use uagents ? Tbh, it looks like to me i *may* be able to build the app with minimal use of urgents or any other fetchai library. Earlier it was clear to me how uagents, agentverse and delta v was absolutely needed, as the solution was being hosted on deltav.
tldr; would a minimum use of uagents in a standalone app be qualified as a valid solution for the problem statement ?

**Reply by abhifetch** [reply]
Is is totally fine if you just have minimal use of uagents in your project. Just try to involve them at the right and most efficient place.

Dont forget to include them. Happy coding

**Reply by saquib000** [reply]
Hi thanks for reply. Quick question, i am getting the binary data from an api call and want to save the file in agentverse envious. So, i am using open() in my agent.py but it says open is not defined. 
Can you please share your thoughts?

**Reply by saquib000** [reply]
Also, i saw earlier integrationa have used many different imports, which are not listed in "allowed imports". So is there a way to get any import? Or has the list of allowed imports have been updated and we can only import these only? 
I basically want to save a file in AV environment, but not sure how to. Any help is appreciated.

**Reply by abhifetch** [reply]
Hi @saquib000 so saving a file is not possible in AV. Also libraries imported in Agentverse are limited.
1. If you want to use agentverse you can use storage. https://fetch.ai/docs/references/uagents/uagents-protocols/storage#agents-storage 
2. If you want to use other libraries can use mailbox and run the agent locally. https://fetch.ai/docs/guides/agents/running-locally#running-locally


I have not tried it but guess if you use locally running agent you can save file as well.

**Comment by AndrewDavidRatnam** [comment]
Could some explain why exactly my schema digest is unrecognized? I'm importing the same Model from the same file in both cases and why is this happening? 

in main.py
from protocols import Input_Action

in subagent..py
from protocols import Input_Action

protocols __init__.py 
def Input_Action(Model):
....

**Reply by abhifetch** [reply]
need more details on the code, if needed join the doubt session today. I will help you resolve this

**Comment by 23f1002382** [comment]
Why was i asked by the DevTeam to download suspicious code?
From the Discord Server?
![Screenshot 2024-05-25 123400](https://github.com/fetchai/uAgents/assets/155766692/0e37b61d-b6df-4278-a4a4-4d464acab1fb)
![Screenshot 2024-05-25 123413](https://github.com/fetchai/uAgents/assets/155766692/7cfc3fce-2344-4ba4-9cdd-d18bd9e56b1b)
 
![Screenshot 2024-05-25 123658](https://github.com/fetchai/uAgents/assets/155766692/a4f8c26e-94bd-45b6-b567-5431bfd1c183)

@abhifetch

**Reply by 23f1002382** [reply]
Ok cheers. This framework is not for me, needs a lot more development from standalone side, which was my hackathon idea. Cheers, Im quitting this Hackathon All the best. May the best man win.

**Reply by abhifetch** [reply]
thanks for joining.

**Comment by nk-droid** [comment]
My agents are not able to get funds. Previously, it was working fine and I haven't changed anything in my code.

![image](https://github.com/fetchai/uAgents/assets/96063103/b670c363-8c4d-4175-8e56-70db0817599f)

@abhifetch

**Reply by abhifetch** [reply]
Can you please update a seed a bit and check again. If not check
Are you using fund_agent_if_low is there or not.

**Reply by nk-droid** [reply]
I already tried that... didn't work. Yes, I'm using fund_agent_if_low().

![image](https://github.com/fetchai/uAgents/assets/96063103/62225cd2-7e68-443e-bebf-dd8bb8c98e7f)

**Reply by abhifetch** [reply]
We are having doubt clearing session tomorrow. Dont worry about deadline, i need to check your code over screen share and then resolve tue issue.

Cheers.

**Reply by saquib000** [reply]
Hi, so like you said now I am now running my agents locally.  But unable to add funds to my agent. 
Checked the docs and some examples and i could find is that already registered agents are able to fetch their fund but not a newly registered agents. 

- As an example from the docs, the below agent is already there on av, and works fine. 
![Screenshot (21)](https://github.com/fetchai/uAgents/assets/103306537/406d38d7-1698-4e90-b688-939a4353f9b2)


- just changing the seed of the agent, registering a new address, makes it not be able to add fund to the agent. 
![Screenshot (22)](https://github.com/fetchai/uAgents/assets/103306537/ff9e58ba-7ea0-45a5-80ce-4053c922a6aa)

@abhifetch 

Sorry for tagging in many threads. appreciate the help
Thanks

**Reply by nk-droid** [reply]
@abhifetch sir, is it possible to share your email id? I can share my code base in zip... You can check and suggest the modifications today only.

**Reply by abhifetch** [reply]
Abhi.gangani@fetch.ai

**Reply by nk-droid** [reply]
I have sent the codebase. Kindly check and suggest modifications.

**Reply by saquib000** [reply]
@nk-droid any update?

**Reply by abhifetch** [reply]
Hello @saquib000 @nk-droid can you please check now if you are able to register agent to alamanc and run fund-agent-if-low?

**Reply by nk-droid** [reply]
Yes, working fine now. Thanks!!!

**Comment by saquib000** [comment]
![20240525_185146](https://github.com/fetchai/uAgents/assets/103306537/da394998-beb7-4abf-b1b3-68a0e4422039)

The function is working on delta v tho. Yet its showing this error on agent verse. 
@abhifetch

**Reply by abhifetch** [reply]
I suggest you to duplicate this function and delete this one.
Get back if this works or not.

**Comment by mdabucse** [comment]
![image](https://github.com/fetchai/uAgents/assets/131733014/bbd4693e-3e87-4dc8-bcd2-119b9e5cc658)
How to resolve this error

**Reply by abhifetch** [reply]
Is your agent in running state and also is it deltaV compatible. i.e. it has protocol and sending messages with UAgentResponse data model.

**Comment by saquib000** [comment]
Unable to create function! 

So I have added mailbox too and yet I am still not able to create functions for my agent. I am assuming something must be wrong with my code, so I just tried to ran code from doc - [https://fetch.ai/docs/guides/agents/locally-running-langchain-integration#running-a-locally-hosted-agent-with-langchain-integration](url) 
Same as my agent, I am able to get it running locally and accessible from AV But not able to create function.
![image](https://github.com/fetchai/uAgents/assets/103306537/63ce762f-ca53-4849-bd26-4b4be1cd3f9d)


Here is my code. just copied from the docs example.

Please take a look at this. @abhifetch 

```from langchain_community.tools import WikipediaQueryRun
from langchain_community.utilities import WikipediaAPIWrapper
from uagents.setup import fund_agent_if_low
from uagents import Agent, Context, Protocol, Model
from pydantic import Field
from ai_engine import UAgentResponse, UAgentResponseType
 
# Extend your protocol with Wikipedia data fetching
class WikiReq(Model):
    search_keyword: str = Field(description="This describes the keyword you want to search on wiki")
 
SEED_PHRASE = "724hqdjq24fjj34"
 
# Copy the address shown below
print(f"Your agent's address is: {Agent(seed=SEED_PHRASE).address}")
 
AGENT_MAILBOX_KEY = "removed the mailbox id for this comment"
 
# Now your agent is ready to join the agentverse!
WikiAgent = Agent(
    name="Wiki Agent",
    seed=SEED_PHRASE,
    mailbox=f"{AGENT_MAILBOX_KEY}@https://agentverse.ai",
)
fund_agent_if_low(WikiAgent.wallet.address()) #funding agent.
wiki_protocol = Protocol("Wiki Protocol")
 
@wiki_protocol.on_message(model=WikiReq, replies={UAgentResponse})
async def load_dalle(ctx: Context, sender: str, msg: WikiReq):
    wikipedia = WikipediaQueryRun(api_wrapper=WikipediaAPIWrapper())
    ctx.logger.info(msg.search_keyword)
    try:
        result = wikipedia.run(msg.search_keyword)
    except Exception as e:
        ctx.logger.info(f"Error generating response: {e}")
    # Send an error response back to the user
    await ctx.send(
        sender, UAgentResponse(message=str(result), type=UAgentResponseType.FINAL)
    )
 
WikiAgent.include(wiki_protocol, publish_manifest=True)
WikiAgent.run()```

**Reply by abhifetch** [reply]
have you replaced the mailbox key and re ran the code before creating function.

**Reply by saquib000** [reply]
Yes I think i have. Once the mailbox key was generated i put it above and re ran the code. It accepted authorization too. Yet the function creation just does not seem to work...

**Reply by mdabucse** [reply]
How to connect the Agent to any one of the frontend

**Reply by abhifetch** [reply]
@mdabucse please refer this repo on how to connect front-end https://github.com/abhifetch/RecipeApp/tree/main.
you have to use on_query decorator with agent. https://fetch.ai/docs/examples/on-query-proxy can be refered as on_query guide.

**Reply by abhifetch** [reply]
@saquib000 mostly your agents quota is over. try to delete unused agents.
create new local agent by changing seed a bit. ideally it should work

**Comment by Nandeesh3000** [comment]
Hello Abhimanyu,

As discussed on the Whatsapp....
![IITM FetchAI](https://github.com/fetchai/uAgents/assets/113001762/5ea4f4cf-c9da-406d-bc35-c4e143d68a82)


Thanks for your inputs. I tried integrating your code snippet into my code but wasn't able to get the output. So, I need your assistance on integrating your code hosted on GitHub with my application. Can we get into the call and have a screen share session to speed up the process?

Thanks

**Reply by abhifetch** [reply]
as discussed over chat please follow https://github.com/abhifetch/RecipeApp/blob/main/Functions%20DeltaV%20Example.pdf 
example to use it in your use case.

**Comment by mdabucse** [comment]
Where we upload our project ?

**Reply by abhifetch** [reply]
https://forms.gle/qxLzb8VVABK7EUKm8

**Reply by abhifetch** [reply]
In case of deltaV integration, make a PR to https://github.com/fetchai/uAgents repo's integration folder.
in case of custom frontend please provide your github repo link and working hosted link for you application

**Comment by nk-droid** [comment]
@abhifetch The [broadcast example](https://fetch.ai/docs/guides/agents/broadcast) given in docs is not working.

I'm getting the following error after copy pasting and running the exact code.
![image](https://github.com/fetchai/uAgents/assets/96063103/6d2c1894-b128-475e-8fb5-8b778d89c11f)

**Reply by abhifetch** [reply]
@nk-droid let me check this. will get back soon.

**Reply by nk-droid** [reply]
@abhifetch any update on this?

**Reply by abhifetch** [reply]
The example is broken. Need to check with development team on this.
Can you please tell me in what way you are trying to use broadcast, so we can find a way around that.

**Reply by nk-droid** [reply]
I'm trying to use the broadcast feature in the same way as shown in the example. There is one agent, and it is calling another agent with the broadcast feature at regular intervals.

**Reply by mdabucse** [reply]
![IMG-20240529-WA0006.jpg](https://github.com/fetchai/uAgents/assets/131733014/43a4ce52-174c-4139-9934-846e12670270)

How to resolve this error

**Reply by abhifetch** [reply]
@nk-droid you can use on_interval
In same way and instead of broadcast use ctx.send to send message.

The reciever agent should have on_message to listen to the recieved message with correct data model.

**Reply by abhifetch** [reply]
# Contributing to uAgents

This guide will help you set up an SSH key for GitHub, clone the `uAgents` repository, create a new branch, and submit a pull request.

## Generating an SSH Key

### macOS

1. **Open Terminal**.
2. **Generate a new SSH key**:
   ```bash
   ssh-keygen -t ed25519 -C "your_email@example.com"
   ```
   If you are using an older system that doesn’t support `ed25519`, use:
   ```bash
   ssh-keygen -t rsa -b 4096 -C "your_email@example.com"
   ```
3. **Specify a file to save the key** (`/Users/you/.ssh/id_ed25519` is the default). Press Enter to accept the default.
4. Optionally, you can add a passphrase for an additional layer of security.

#### Adding the SSH Key to the SSH-Agent

1. **Start the SSH agent in the background**:
   ```bash
   eval "$(ssh-agent -s)"
   ```
2. **Add your SSH key to the SSH agent**:
   ```bash
   ssh-add ~/.ssh/id_ed25519
   ```

#### Adding the SSH Key to Your GitHub Account

1. **Copy the SSH key to your clipboard**:
   ```bash
   pbcopy < ~/.ssh/id_ed25519.pub
   ```
2. **Go to GitHub and log in**.
3. Navigate to **Settings** > **SSH and GPG keys** > **New SSH key**.

![Screenshot 2024-05-29 at 08 38 40](https://github.com/fetchai/uAgents/assets/156676541/5e69782f-fff8-4aa5-80f8-5efd3b2c5ccb)

4. **Paste your key and give it a descriptive title**.


![Screenshot 2024-05-29 at 08 39 07](https://github.com/fetchai/uAgents/assets/156676541/774ab14a-7583-4362-ba7a-820cd58ed1ac)

5. Click add ssh key

### Windows

1. **Open Git Bash**.
2. **Generate a new SSH key**:
   ```bash
   ssh-keygen -t ed25519 -C "your_email@example.com"
   ```
   For older systems:
   ```bash
   ssh-keygen -t rsa -b 4096 -C "your_email@example.com"
   ```
3. **Follow the prompts to save the key and optionally add a passphrase**.

#### Adding the SSH Key to the SSH-Agent

1. **Start the SSH agent in the background**:
   ```bash
   eval "$(ssh-agent -s)"
   ```
2. **Add your SSH key to the SSH agent**:
   ```bash
   ssh-add ~/.ssh/id_ed25519
   ```

#### Adding the SSH Key to Your GitHub Account

1. **Copy the SSH key to your clipboard**:
   ```bash
   clip < ~/.ssh/id_ed25519.pub
   ```
2. **Go to GitHub and log in**.
3. Navigate to **Settings** > **SSH and GPG keys** > **New SSH key**.
4. **Paste your key and give it a descriptive title**.

## Cloning the Repository and Creating a Branch

1. **Clone the Repository**:
   ```bash
   git clone git@github.com:fetchai/uAgents.git
   cd uAgents
   ```

2. **Create a New Branch**:
   Replace `your-branch-name` with a descriptive name for your branch:
   ```bash
   git checkout -b your-branch-name
   ```

3. **Make Changes and Commit**:
   Make your changes to the codebase, then add and commit them:
  You have to add your folder in integration folder

   ```bash
   git add .
   git commit -m "Description of the changes you made"
   ```

4. **Push the Branch to GitHub**:
   ```bash
   git push -u origin your-branch-name
   ```

## Creating a Pull Request

1. **Go to the repository on GitHub**.
2. You will see a prompt to **compare & pull request** for your recently pushed branch.
3. Click **Compare & pull request**.
4. **Fill in the details for your pull request and create it**.

That's it! You've created an SSH key, cloned a repository, created a branch, pushed your changes, and created a pull request.

**Comment by serdonke** [comment]
Hi, I have a python script that runs a flask server and serves a page and I have a uagent script that communicates with an Agent using uagents, I want to take the data from the server submitted via a form and then send it to my uagents script for further processing but I am not sure how to do that since uagents uses async functions and such.

**Reply by abhifetch** [reply]
You have to query your agent using uagents.query.
in your agent you have to use 'on_query' handler to process query.

sample to use this provided in this repo. (check backend) https://github.com/abhifetch/RecipeApp
